File: packages/capacitor-geofence/android/src/main/java/com/sweptmind/geofence/GeofencePlugin.kt
package com.sweptmind.geofence

import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "Geofence")
class GeofencePlugin : Plugin() {
    private lateinit var geofenceManager: GeofenceManager

    override fun load() {
        geofenceManager = GeofenceManager(context)
        geofenceManager.onTransition = { identifier, type, latitude, longitude ->
            val data = JSObject().apply {
                put("identifier", identifier)
                put("type", type)
                put("latitude", latitude)
                put("longitude", longitude)
            }
            notifyListeners("geofenceTransition", data)
        }
    }

    @PluginMethod
    fun addGeofences(call: PluginCall) {
        val geofencesArray = call.getArray("geofences") ?: run {
            call.reject("Missing geofences array")
            return
        }

        val configs = mutableListOf<GeofenceConfig>()
        for (i in 0 until geofencesArray.length()) {
            val obj = geofencesArray.getJSONObject(i)
            configs.add(
                GeofenceConfig(
                    identifier = obj.getString("identifier"),
                    latitude = obj.getDouble("latitude"),
                    longitude = obj.getDouble("longitude"),
                    radiusMeters = obj.getDouble("radiusMeters").toFloat(),
                    notifyOnEntry = obj.optBoolean("notifyOnEntry", true),
                    notifyOnExit = obj.optBoolean("notifyOnExit", false),
                    notificationTitle = obj.optString("notificationTitle", null),
                    notificationBody = obj.optString("notificationBody", null),
                )
            )
        }

        geofenceManager.addGeofences(configs)
        call.resolve()
    }

    @PluginMethod
    fun removeGeofences(call: PluginCall) {
        val identifiers = call.getArray("identifiers") ?: run {
            call.reject("Missing identifiers array")
            return
        }
        val ids = mutableListOf<String>()
        for (i in 0 until identifiers.length()) {
            ids.add(identifiers.getString(i))
        }
        geofenceManager.removeGeofences(ids)
        call.resolve()
    }

    @PluginMethod
    fun removeAllGeofences(call: PluginCall) {
        geofenceManager.removeAllGeofences()
        call.resolve()
    }

    @PluginMethod
    fun getMonitoredGeofences(call: PluginCall) {
        val geofences = geofenceManager.getMonitoredGeofences()
        val result = JSArray()
        for (g in geofences) {
            result.put(JSObject().apply {
                put("identifier", g.identifier)
                put("latitude", g.latitude)
                put("longitude", g.longitude)
                put("radiusMeters", g.radiusMeters)
            })
        }
        call.resolve(JSObject().apply { put("geofences", result) })
    }

    @PluginMethod
    fun requestAlwaysPermission(call: PluginCall) {
        // On Android, background location is requested via runtime permissions
        // which Capacitor handles. For now, check the current status.
        call.resolve(JSObject().apply {
            put("status", geofenceManager.getPermissionStatus())
        })
    }

    @PluginMethod
    fun getPermissionStatus(call: PluginCall) {
        call.resolve(JSObject().apply {
            put("status", geofenceManager.getPermissionStatus())
        })
    }
}

File: packages/capacitor-geofence/ios/Sources/GeofencePlugin/GeofencePlugin.swift
import Capacitor
import CoreLocation

@objc(GeofencePlugin)
public class GeofencePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "GeofencePlugin"
    public let jsName = "Geofence"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "addGeofences", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeGeofences", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeAllGeofences", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getMonitoredGeofences", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestAlwaysPermission", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPermissionStatus", returnType: CAPPluginReturnPromise),
    ]

    private lazy var manager: GeofenceManager = {
        let mgr = GeofenceManager()
        mgr.onTransition = { [weak self] identifier, type, latitude, longitude in
            self?.notifyListeners("geofenceTransition", data: [
                "identifier": identifier,
                "type": type,
                "latitude": latitude,
                "longitude": longitude,
            ])
        }
        return mgr
    }()

    private var permissionCall: CAPPluginCall?

    @objc func addGeofences(_ call: CAPPluginCall) {
        guard let geofencesArray = call.getArray("geofences") as? [JSObject] else {
            call.reject("Missing geofences array")
            return
        }

        var configs: [(identifier: String, latitude: Double, longitude: Double, radiusMeters: Double, notifyOnEntry: Bool, notifyOnExit: Bool, notificationTitle: String?, notificationBody: String?)] = []

        for obj in geofencesArray {
            guard let identifier = obj["identifier"] as? String,
                  let latitude = obj["latitude"] as? Double,
                  let longitude = obj["longitude"] as? Double,
                  let radiusMeters = obj["radiusMeters"] as? Double else {
                continue
            }
            let notifyOnEntry = obj["notifyOnEntry"] as? Bool ?? true
            let notifyOnExit = obj["notifyOnExit"] as? Bool ?? false
            let notificationTitle = obj["notificationTitle"] as? String
            let notificationBody = obj["notificationBody"] as? String

            configs.append((identifier, latitude, longitude, radiusMeters, notifyOnEntry, notifyOnExit, notificationTitle, notificationBody))
        }

        manager.addGeofences(configs)
        call.resolve()
    }

    @objc func removeGeofences(_ call: CAPPluginCall) {
        guard let identifiers = call.getArray("identifiers") as? [String] else {
            call.reject("Missing identifiers array")
            return
        }
        manager.removeGeofences(identifiers: identifiers)
        call.resolve()
    }

    @objc func removeAllGeofences(_ call: CAPPluginCall) {
        manager.removeAllGeofences()
        call.resolve()
    }

    @objc func getMonitoredGeofences(_ call: CAPPluginCall) {
        let geofences = manager.getMonitoredGeofences()
        call.resolve(["geofences": geofences])
    }

    @objc func requestAlwaysPermission(_ call: CAPPluginCall) {
        permissionCall = call
        manager.requestAlwaysPermission { [weak self] status in
            self?.permissionCall?.resolve(["status": status])
            self?.permissionCall = nil
        }
    }

    @objc func getPermissionStatus(_ call: CAPPluginCall) {
        call.resolve(["status": manager.getPermissionStatus()])
    }
}
